### PR TITLE
reduceWhile fixes #1737

### LINKED ISSUE #1737: reduceWhile
any interest in `reduceWhile`? This is simply sugar for `reduce` using a passed-in predicate to detremine whether to short-circuit (i.e. `reduced(acc)`) or keep evaluating. I think it might be nicer than the current somewhat awkward use of `reduced`.

Also, should the predicate be `x -> Bool` or `(acc, x) -> Bool`? I have my opinion; I'm curious if anyone sees reasons to favor one or the other.

### PR DESCRIPTION
**Pull request by buzzdecafe**
Closes #1737

### CODE REVIEW

File: src/reduceWhile.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * Like `reduce`, `reduceWhile` returns a single item by iterating through
+ * the list, successively calling the iterator function. `reduceWhile` also
+ * takes a predicate that is evaluated at each step. If the predicate returns
+ * `false`, it "short-circuits" the iteration and returns the current value
+ * of the accumulator.
+ *
+ * @func
+ * @memberOf R
+ * @category List
+ * @sig ((a, b) -> Boolean) -> (a, b) -> a) -> a -> [b] -> a
```
Comment (davidchambers) [inline src/reduceWhile.js]:
```diff
- * @sig ((a, b) -> Boolean) -> (a, b) -> a) -> a -> [b] -> a
+ * @sig ((a, b) -> Boolean) -> ((a, b) -> a)) -> a -> [b] -> a
```

Comment (buzzdecafe) [inline src/reduceWhile.js]: well spotted as usual

File: src/reduceWhile.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+var _curryN = require('./internal/_curryN');
+var _reduce = require('./internal/_reduce');
+var _reduced = require('./internal/_reduced');
+
+
+/**
+ * Like `reduce`, `reduceWhile` returns a single item by iterating through
+ * the list, successively calling the iterator function. `reduceWhile` also
+ * takes a predicate that is evaluated at each step. If the predicate returns
```
Comment (asaf-romano) [inline src/reduceWhile.js]: The word "before" is missing here, somewhere. It should be clear that the predicate is running before the reducer (and by extension, that the accumulator may be returned as-is).

File: src/reduceWhile.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * @param {*} a The accumulator value.
+ * @param {Array} list The list to iterate over.
+ * @return {*} The final, accumulated value.
+ * @see R.reduce, R.reduced
+ * @example
+ *
+ *      var isOdd = (x) => x mod 2 === 1;
+ *      var xs = [1, 3, 5, 60, 777, 800];
+ *
+ *      R.reduceWhile(isOdd, R.add, 0, xs); //=> 9
```
Comment (asaf-romano) [inline src/reduceWhile.js]: It would be nice to also include an initial-accumulator-returned example.

File: src/reduceWhile.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ *        current element.
+ * @param {Function} fn The iterator function. Receives two values, the
+ *        accumulator and the current element.
+ * @param {*} a The accumulator value.
+ * @param {Array} list The list to iterate over.
+ * @return {*} The final, accumulated value.
+ * @see R.reduce, R.reduced
+ * @example
+ *
+ *      var isOdd = (x) => x mod 2 === 1;
```
Comment (benperez) [inline src/reduceWhile.js]: it might be more clear if you write this as:
```javascript
var isOdd = x => x % 2 === 1;
```

Comment (CrossEye) [inline src/reduceWhile.js]: But isn't this still the wrong function?

Shouldn't it be 

```js
    (acc, x) => x % 2 === 1
```

??

Comment (buzzdecafe) [inline src/reduceWhile.js]: yeah, too much "not js" lately. By which I mean, not enough.

File: src/reduceWhile.js (hunk 2 above)
Comment (buzzdecafe) [inline src/reduceWhile.js]: :heavy_check_mark:

File: src/reduceWhile.js (hunk 3 above)
Comment (buzzdecafe) [inline src/reduceWhile.js]: :heavy_check_mark:

File: src/reduceWhile.js (hunk 4 above)
Comment (buzzdecafe) [inline src/reduceWhile.js]: :heavy_check_mark:

**Comment by asaf-romano** [comment]
🌷

File: src/reduceWhile.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * @return {*} The final, accumulated value.
+ * @see R.reduce, R.reduced
+ * @example
+ *
+ *      var isOdd = (acc, x) => x % 2 === 1;
+ *      var xs = [1, 3, 5, 60, 777, 800];
+ *      R.reduceWhile(isOdd, R.add, 0, xs); //=> 9
+ *
+ *      var ys = [2, 4, 6]
+ *      R.reduceWhile(isOdd, R.add, 111, xs); //=> 111
```
Comment (kedashoe) [inline src/reduceWhile.js]: pass in `ys` rather than `xs` here?

I :heart_decoration: the simple examples

Comment (buzzdecafe) [inline src/reduceWhile.js]: :zap:

**Comment by CrossEye** [comment]
:herb:

**Pull request merged**